jena-shacl/src/main/java/org/apache/jena/shacl/lib/ShLib.java

```diff
@@ -195,6 +208,25 @@ public static boolean isImmediate(Target target) {
         return targetType.equals(TargetType.targetObjectsOf) || targetType.equals(TargetType.targetSubjectsOf);
     }
 
+    static Set<String> rdfDatatypes = new HashSet<>();
+    static {
+        rdfDatatypes.add(RDF.dtLangString.getURI());
+        rdfDatatypes.add(RDF.dtRDFHTML.getURI());
+        rdfDatatypes.add(RDF.dtRDFJSON.getURI());
+        rdfDatatypes.add(RDF.dtXMLLiteral.getURI());
+    }
+
+
+
+    /** Test whether the IRI is a datatype that can be written in compact short form (no datatype=)
```

Comment:
"datatype=" intended. 

A plain URI as a constraint is a shorthand for either "datatype=URI" or "class=URI" by looking inside the URI.  Teh writer attempts to reverse this conservatively. For XSD and some known datatypes, it writes the URI as an abbreviation for datatype=URI otherwise it writes class= (no attempt to abbreviate)

I'll clarify the comment.